### PR TITLE
Move the reference sequence number to the runtime

### PR DESCRIPTION
**Pull request by andre4i**
## Description

Related to ADO:3397

There is a known issue regarding reference sequence number mismatches. It is possible that the inbound queue is scheduled to start processing ops before the flush (which creates a batch) is scheduled. This may create the situation in which an op is created after the inboud queue processes, and the op will have a different reference sequence number than the rest of the batch.

Currently, the loader does not care about this, as it is not receiving the reference sequence number from the runtime at all. This will still be the case with old loaders.

This change does the following:

- moves the reference sequence number stamping at the batch level in the runtime
- it guarantees that all batches will have the same reference sequence numbers by flushing the partial batches whenever we detect a sequence number mismatch (unless the feature is disabled by feature gate). The change is correct from the perspective of batching, as if the deltamanager started processing ops, it is clear that the current batch should be over, as the current JS turn is over as well.
- this change is required for both compression/chunking correctness and for the [experimental flushmode async](https://github.com/microsoft/FluidFramework/pull/14106), which may create more situations like described above.

### CODE REVIEW

File: packages/loader/container-loader/src/container.ts [hunk 1]
```diff
@@ -1964,6 +1964,7 @@ export class Container
 				true, // batch
 				message.metadata,
 				message.compression,
+				message.referenceSequenceNumber,
```
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:1967]: I'm not sure if we want to allow different ref seq number for different ops in a batch. I.e. should it be a property of op, or a batch?
It feels like (with compression & possibly chunking requirements) we are settling on enforcing at runtime level that all ops in a batch should have same ref seq numbers. If so, we might want to pass ref seq number as an additional argument to submitBatch() API.

File: packages/loader/container-loader/src/container.ts [hunk 2]
```diff
@@ -1992,6 +1993,7 @@ export class Container
 		batch?: boolean,
 		metadata?: any,
 		compression?: string,
+		referenceSequenceNumber?: number,
```
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:1996]: I'd not worry about this path - that's legacy path used by old runtimes, that will not provide this info any way, so there is not much point to add it here.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 3]
```diff
@@ -2843,7 +2845,7 @@ export class ContainerRuntime
 			deserializedContent: JSON.parse(serializedContent), // Deep copy in case caller changes reference object
 			metadata,
 			localOpMetadata,
-			referenceSequenceNumber: this.deltaManager.lastSequenceNumber,
+			referenceSequenceNumber: this.lastProcessedSequenceNumber,
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2848]: I do not think having this.lastProcessedSequenceNumber buys us anything. Old code does as good.
What is important (in this PR) is that we pass this info to Loader layer - we were already capturing right data here, just not passing it through.

File: packages/loader/container-loader/src/container.ts (hunk 2 above)
Comment (andre4i) [reply packages/loader/container-loader/src/container.ts:1996]: Not sure if it's the legacy path. This is used by `private submitBatch(batch: IBatchMessage[]): number` above.

Comment (vladsud) [reply packages/loader/container-loader/src/container.ts:1996]: Correct - it's feature detection. Runtime will try to use submitBatchFn() if it exists, otherwise switch to submitFn()
The version of runtime that has support for refSeqNumber will have submitBatchFn(), so runtime will chose it.
The exception might be for some stand-alone ops (like summarize Op), but I'm not sure it matters.
Double check these claims, please :)

Comment (andre4i) [reply packages/loader/container-loader/src/container.ts:1996]: That's correct. And if we want to give the referenceSequenceNumber to the loader, we will need it in the signature of this method, as it is used by submitBatchFn and the latter is the entry-point from the runtime.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 4]
```diff
@@ -1427,6 +1433,18 @@ export class ContainerRuntime
...
+				attachReorderDisabled: this.mc.config.getBoolean(
+					"Fluid.ContainerRuntime.disableAttachOpReorder",
+				),
+				partialFlushDisabled: this.mc.config.getBoolean(
+					"Fluid.ContainerRuntime.DisablePartialFlush",
+				),
+			}),
 		});
 
 		ReportOpPerfTelemetry(this.context.clientId, this.deltaManager, this.logger);
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1450]: changes in this file seems unrelated to the core of PR, I'd suggest pushing them separately.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts [hunk 5]
```diff
@@ -32,40 +36,62 @@ export interface IOutboxParameters {
...
+			this.mc.logger.sendErrorEvent(
+				{
+					eventName: "ReferenceSequenceNumberMismatch",
+					referenceSequenceNumber: batchReference,
+					messageReferenceSequenceNumber: message.referenceSequenceNumber,
+					type: message.deserializedContent.type,
+					count: batchManager.length,
+					batchSize: batchManager.contentSizeInBytes,
+				},
+				new UsageError("Submission of an out of order message"),
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/opLifecycle/outbox.ts:82]: I'd not add error object here - it has no purpose. Event name is all we care about.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts [hunk 6]
```diff
@@ -32,40 +36,62 @@ export interface IOutboxParameters {
 }
 
 export class Outbox {
+	private readonly mc: MonitoringContext;
 	private readonly attachFlowBatch: BatchManager;
 	private readonly mainBatch: BatchManager;
-	private readonly defaultAttachFlowSoftLimitInBytes = 64 * 1024;
+	private readonly defaultAttachFlowSoftLimitInBytes = 320 * 1024;
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/opLifecycle/outbox.ts:42]: This number is the limit before compression, right? It's very hard to say what algorithm we should follow to find "best" number here. It feels like if app is using timer-based flush, they tell us - they do not care about latency, but care much more about batch counts. And if so, we might use this hint to substantially increase this limit, or maybe even - eliminate attach ops sending ahead of time (as grouping all ops into smaller set of batches will result in better througtput / lower cost for consumers, at the expense of higher latency.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts (hunk 5 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:82]: it was not really useful in production but during testing, the stack trace from this error definitely helps.

File: packages/runtime/container-runtime/src/test/opLifecycle/batchManager.spec.ts [hunk 7]
```diff
@@ -165,73 +159,12 @@ describe("BatchManager", () => {
...
 		assert.equal(batchManager.contentSizeInBytes, smallMessageSize * batchManager.length);
 		assert.equal(batchManager.push(smallMessage()), true);
 		assert.equal(batchManager.contentSizeInBytes, smallMessageSize * batchManager.length);
 	});
-
-	it("Don't verify op ordering by default", () => {
-		const batchManager = new BatchManager({ hardLimit }, mockLogger);
-		assert.equal(batchManager.push({ ...smallMessage(), referenceSequenceNumber: 0 }), true);
-		assert.equal(batchManager.push({ ...smallMessage(), referenceSequenceNumber: 0 }), true);
-		assert.equal(batchManager.push({ ...smallMessage(), referenceSequenceNumber: 1 }), true);
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/test/opLifecycle/batchManager.spec.ts:181]: It feels like we still need these tests, it's just that we should test for different outcome, no?

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts (hunk 6 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:42]: > This number is the limit before compression, right? 

Yes. I think we discussed this being too aggressive.

File: packages/runtime/container-runtime/src/test/opLifecycle/batchManager.spec.ts (hunk 7 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/test/opLifecycle/batchManager.spec.ts:181]: The batch manager is no longer governing this. The tests will be added to the outbox.

File: packages/loader/container-loader/src/container.ts (hunk 1 above)
Comment (andre4i) [reply packages/loader/container-loader/src/container.ts:1967]: working on it

File: packages/loader/container-loader/src/container.ts [hunk 8]
```diff
@@ -1992,6 +1993,7 @@ export class Container
 		batch?: boolean,
```
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:1993]: It feels like submitSummaryMessage() should be more explicit and pass ref seq number to submitMessage().
We do not have any better info (I thought seq number would be in ISummaryContent, but there is nothing there).
So maybe add optional argument to submitSummaryMessage() API such that runtime (in future) could supply ref sequence number, and thus (for example) produce a summary (from old state) while inbounding ops. Today that's not possible as our DDSs do not have capability to summaize old state (and exclude local changes), but maybe in future we will have that capability, and summarizer code will need to supply ref seq number.

**Review by vladsud — COMMENTED** [review]
Overall looks good. But I'd rather (at least at protocol level) make ref seq number a property of whole batch, and not pass it on per-op granularity.

**Comment by andre4i** [comment]
> Overall looks good. But I'd rather (at least at protocol level) make ref seq number a property of whole batch, and not pass it on per-op granularity.

@vladsud 
thank you! Working on specifically that issue right now.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 9]
```diff
@@ -1290,7 +1290,9 @@ export class ContainerRuntime
 			config: {
 				compressionOptions,
 				maxBatchSizeInBytes: runtimeOptions.maxBatchSizeInBytes,
-				enableOpReentryCheck: this.enableOpReentryCheck,
+				disablePartialFlush:
+					this.mc.config.getBoolean("Fluid.ContainerRuntime.DisablePartialFlush") ===
+					true,
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1295]: unrelated? Seems to wipe out enableOpReentryCheck

Comment (andre4i) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1295]: For the outbox, this is intentional. We won't be checking for op reentry. I mean we'll always do op reentry check but instead of closing the container, we log the instance and (if not disabled) split the batch.

`enableOpReentryCheck` is still in effect for blocking ops from within event handlers. That's caught earlier than attempting to add the op to the batch (via the event handler API on the DDS which will call into the runtime's ensureNoDataModelChanges function)

File: packages/loader/container-loader/src/container.ts [hunk 10]
```diff
@@ -1964,6 +1964,7 @@ export class Container
 				true, // batch
 				message.metadata,
 				message.compression,
+				referenceSequenceNumber ?? message.referenceSequenceNumber,
```
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:1967]: This feels questionable to me. The way interface is structured right now - it basically says we are Ok allowing different ref sequence numbers on ops in a batch, but we will also provide an overwrite for whole batch. Runtime, if it wants to, can overwrite ref seq number on every op.
I'd not add IBatchMessage.referenceSequenceNumber, and instead this code use either ref seq number provided by runtime (for whole batch) or default one from delta manager. And in future (next) releases remove capability for later (i.e. make argument required).

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts [hunk 11]
```diff
@@ -186,7 +227,12 @@ export class Outbox {
 					contents: message.contents,
 					metadata: message.metadata,
 					compression: message.compression,
+					referenceSequenceNumber: message.referenceSequenceNumber,
 				})),
+				// Partial flushing will ensure that all messages in the same batch have the same
+				// reference sequence number. If disabled, messages may have different sequence numbers
+				// within the same batch, so the loader must rely on individual message reference sequence numbers.
+				this.params.config.disablePartialFlush ? undefined : batch.referenceSequenceNumber,
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/opLifecycle/outbox.ts:235]: I'd always use batch.referenceSequenceNumber here. We already (before this change) use single ref sequence number for a batch - DeltaManager does that, plus optinal compresison does it (on top of it).
I do not see much value in allowing to go back and re-implement per-message ref sequence number capability - I'd always pass batch.referenceSequenceNumber

**Review by vladsud — APPROVED** [review]

File: packages/loader/container-loader/src/container.ts (hunk 10 above)
Comment (andre4i) [reply packages/loader/container-loader/src/container.ts:1967]: Thanks for calling this out, I had my doubts about creating too many states.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts (hunk 11 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:235]: But if we `disablePartialFlush` is set to true, we won't be splitting the batch in case we find a different reference sequence number. 

This means that the batch will have different sequence numbers and we're going to be lying to the loader about the shape of the whole batch.

We can either:
- remove the `disablePartialFlush` feature gate
- keep this as is and if we supply `undefined` here, the delta manager should use it's own 'lastprocessedseqnum` instead (based on [the comment above](https://github.com/microsoft/FluidFramework/pull/14150#discussion_r1107604283))

Comment (vladsud) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:235]: It's no different than what we have today. I.e. today, DeltaManager will overwrite ref seq number. If we always pass ref seq number (even if disablePartialFlush is true), then it's either noop, or we will supply smaller (earlier) ref seq number, but it will be more correct (assuming that we have newer loader that picks it up).
It's a small change in behavior, so yes, there is some risk, but I think it's tiny.

That said, I think it's beneficial (for time being) to use last op (in a batch) ref seq number when calculating ref sequence number for a batch, not first op. If they are different, it's less risk to use slightly higher number (but possibly lower than what DeltaManager might have stamped today)

Comment (andre4i) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:235]: Thank you, it makes sense. I'll make the change regarding the batch reference sequence number. I agree it's less risk to use the highest ref seq num.

**Comment by vladsud** [comment]
LGTM!

File: packages/loader/container-loader/src/container.ts [hunk 12]
```diff
@@ -2088,7 +2097,8 @@ export class Container
 			(type, contents, batch, metadata) =>
 				this.submitContainerMessage(type, contents, batch, metadata),
 			(summaryOp: ISummaryContent) => this.submitSummaryMessage(summaryOp),
-			(batch: IBatchMessage[]) => this.submitBatch(batch),
+			(batch: IBatchMessage[], referenceSequenceNumber?: number) =>
```
Comment (agarwal-navin) [inline packages/loader/container-loader/src/container.ts:2100]: submitSummaryMessage also needs to be updated to have referenceSequenceNumber.

File: packages/loader/container-loader/src/container.ts [hunk 13]
```diff
@@ -1927,7 +1927,7 @@ export class Container
 	}
 
 	/** @returns clientSequenceNumber of last message in a batch */
-	private submitBatch(batch: IBatchMessage[]): number {
+	private submitBatch(batch: IBatchMessage[], referenceSequenceNumber?: number): number {
```
Comment (agarwal-navin) [inline packages/loader/container-loader/src/container.ts:1930]: submitSummaryMessage below also needs to be updated to have referenceSequenceNumber as a param

File: packages/loader/container-loader/src/deltaManager.ts [hunk 14]
```diff
@@ -230,11 +230,14 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
 		batch = false,
 		metadata?: any,
 		compression?: string,
+		referenceSequenceNumber?: number,
 	) {
+		// Back-compat ADO:3455
```
Comment (agarwal-navin) [inline packages/loader/container-loader/src/deltaManager.ts:235]: I don't think we should be adding reference to ADO items.

File: packages/loader/container-loader/src/deltaManager.ts [hunk 15]
```diff
@@ -230,11 +230,14 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
 		batch = false,
 		metadata?: any,
 		compression?: string,
+		referenceSequenceNumber?: number,
```
Comment (agarwal-navin) [inline packages/loader/container-loader/src/deltaManager.ts:233]: Can we add a submitBatch function that Container calls instead of this? (Maybe in a separate PR)

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts [hunk 16]
```diff
@@ -32,40 +38,74 @@ export interface IOutboxParameters {
...
+		const batchReference = batchManager.referenceSequenceNumber;
+		if (batchReference === undefined || batchReference === message.referenceSequenceNumber) {
+			// The reference sequence numbers are stable, there is nothing to do
+			return;
+		}
+
+		if (++this.mismatchedOpsReported <= this.maxMismatchedOpsToReport) {
+			this.mc.logger.sendErrorEvent(
+				{
+					eventName: "ReferenceSequenceNumberMismatch",
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/opLifecycle/outbox.ts:91]: We know this is possible right so this should not be an error. Are we logging this to understand how often this happens?

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts [hunk 17]
```diff
@@ -77,6 +117,8 @@ export class Outbox {
 	}
 
 	public submitAttach(message: BatchMessage) {
+		this.maybeFlushPartialBatch(this.mainBatch, message);
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/opLifecycle/outbox.ts:120]: Shouldn't this apply to attach message too? What I mean is attach messages batch should also have the same reference sequence number right?

Comment (vladsud) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:120]: It does not matter, but I think we want to get to the point where this extra arg is required, so yes, we should add it to all calls.

File: packages/loader/container-loader/src/deltaManager.ts (hunk 14 above)
Comment (andre4i) [reply packages/loader/container-loader/src/deltaManager.ts:235]: I remember we had a discussion at some point and the consensus was that we shouldn't add links but notes should be fine. I don't think we need to open issues specifically for these things. It's just a TODO with accountability.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts (hunk 16 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:91]: Yes. This was moved from the batch manager.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts (hunk 17 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:120]: Ohhhh... Nice catch.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts (hunk 16 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:91]: I partially agree with you here. I'd say - overall we do not control scheduling of micro-tasks, so these conditions will happen in natural flow of things.
That said, I think that most of these things today happen due to reentrnacy, i.e., sending ops from DDS event handlers (while ops are applied). It would be awesome if we can fully disallow this flow and validate that helps (statistically).

Ideally we do all of that before starting to play with 0ms timer-based flushes....

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:91]: _I think that most of these things today happen due to reentrnacy_

If so, how can we identify if that happens. Right now, this log will not help us distinguish re-entrancy vs real cases where ops are scheduled via Promises.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts (hunk 17 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:120]: @vladsud Why do you say it does not matter? How is attach messages batch different from regular batch?

Comment (andre4i) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:120]: Yeah, this is a legitimate bug. The current code will flush the main batch (only) when it detects an attach op with a different reference sequence number. I think the `maybeFlushPartialBatch` should flush both batches when this happens for either attach ops or regular ops, as the reference sequence number mismatch will be in both.

Adding a unit test for these cases right now.

File: packages/loader/container-loader/src/containerContext.ts [hunk 18]
```diff
@@ -57,8 +57,8 @@ export class ContainerContext implements IContainerContext {
 		quorum: IQuorum,
 		loader: ILoader,
 		submitFn: (type: MessageType, contents: any, batch: boolean, appData: any) => number,
-		submitSummaryFn: (summaryOp: ISummaryContent) => number,
-		submitBatchFn: (batch: IBatchMessage[]) => number,
+		submitSummaryFn: (summaryOp: ISummaryContent, referenceSequenceNumber?: number) => number,
```
Comment (agarwal-navin) [inline packages/loader/container-loader/src/containerContext.ts:60]: ContainerRuntime should add referenceSequenceNumber when sending summary message. That change is missing.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts (hunk 16 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:99]: Oh, I see what you mean, Navin. We can, actually. Check out `verifyCanSubmitOps` in ContainerRuntime. If it's due to op reentry from the DDS event handlers, we will know.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts (hunk 17 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts]: updated @agarwal-navin @vladsud

File: packages/loader/container-loader/src/containerContext.ts (hunk 18 above)
Comment (andre4i) [reply packages/loader/container-loader/src/containerContext.ts:60]: Not sure if we can right now: https://github.com/microsoft/FluidFramework/pull/14150#discussion_r1106452120

I'd rather leave that to use the old flow for now, but have the loader ready to receive it when it's settled.

Comment (agarwal-navin) [reply packages/loader/container-loader/src/containerContext.ts:60]: If I understand correctly, I think Vlad is talking about a future where summaries can happen while inbound ops are flowing. However, today we can still send the referenceSequenceNumber when submitting the summary message. It would be the same as the one delta manager would add so not adding it will have the same behavior, but I think we should still do it for consistency.

Not doing it in this PR is fine, but we at least should track it.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts [hunk 19]
```diff
@@ -32,40 +38,80 @@ export interface IOutboxParameters {
...
+	 * what was already in the batch managers, this means that batching has been interrupted so
+	 * we will flush the accumulated messages to account for that and create a new batch with the new
+	 * message as the first message.
+	 *
+	 * @param message - the incoming message
+	 */
+	private maybeFlushPartialBatch(message: BatchMessage) {
+		const mainBatchReference = this.mainBatch.referenceSequenceNumber;
+		const attachFlowBatchReference = this.attachFlowBatch.referenceSequenceNumber;
+		assert(
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/opLifecycle/outbox.ts:83]: Are we confident that this assert won't hit? I am asking because this has known to happen :). And if so, starting with an error would give us more info if it happens.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts [hunk 20]
```diff
@@ -32,40 +38,80 @@ export interface IOutboxParameters {
...
+			mainBatchReference === undefined ||
+				attachFlowBatchReference === undefined ||
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/opLifecycle/outbox.ts:85]: lol prettier formatting for this.

**Review by agarwal-navin — APPROVED** [review]

**Comment by agarwal-navin** [comment]
@andre4i Can you please add a descriptive title and decription since this is not a draft PR anymoe :-)

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts (hunk 19 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:83]: I'm working on refining this. It won't be hit as long as we don't disable partial flushes.

File: packages/loader/container-loader/src/containerContext.ts (hunk 18 above)
Comment (andre4i) [reply packages/loader/container-loader/src/containerContext.ts:60]: ah, I see. Yeah, it's an easy change in this case.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts (hunk 17 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts]: What I meant - ref seq number of attach op does not matter, as result of applying it does not matter when client send it (what state it had when sending it)

Comment (vladsud) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts]: BTW, we should reconsider having separate batch for attach ops. It was more of a short term solution to reduce batch sizes. With compression and chunking the value of this feature is lower. It gives us maybe some benefit on latency (ops go out sooner and applied sooner), but it also gives us drawback in terms of lower compression rates and more (overall) messages being sent (that will mean more ops once we have seq number virtualization). I'm not sure benefits outweigh cons (including more code / more complexity), and we do not have good way to measure  impact here, so I think removing all this logic is fine.

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts]: I like the idea of removing the separate batch for attach ops. It adds additions flows for sending batches which increases complexity and thereby changes of things going wrong (this issue that was luckily caught and fixed for example).

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts [hunk 21]
```diff
@@ -32,40 +38,86 @@ export interface IOutboxParameters {
...
+	private readonly defaultAttachFlowSoftLimitInBytes = 320 * 1024;
+
+	/**
+	 * Track the number of ops which were detected to have a mismatched
+	 * reference sequence number, in order to self-throttle the telemetry events.
+	 *
+	 * This should be removed as part of ADO:2322
+	 */
+	private readonly maxMismatchedOpsToReport = 3;
+	private mismatchedOpsReported = 0;
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/opLifecycle/outbox.ts:53]: You can get away with just one counter that decrements to zero

Comment (andre4i) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:53]: I know, but I want the `x` out of `y` in the logs. We could deduce that with crafting a better kusto query but I figured the bundle size hit is minimal.

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts (hunk 17 above)
Comment (andre4i) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts]: Yeah, we will do that after compression/chunking goes mainstream (opt-out, not opt-in)
